### PR TITLE
[Core][plasma] fix the data race issue

### PR DESCRIPTION
**Pull request by scv119**
TL;DR: ProcessCreateRequests will be [scheduled asynchronously](https://github.com/ray-project/ray/pull/18312/files#diff-5698038aacf457a9c1e23a09a422a1fdc8dec111fd59929de955feffc150b075R678) when the initial call failed. Unfortunately this is not protected by the lock and could cause data race. 

More details:

-  IsObjectSpillable is called by a different thread from LocalObjectManager and protected by a PlasmaStore.cc mutex;
- ProcessCreateRequests is called by PlasmaStore.cc HandleRequest in PlasmaStore’s own thread. The HandleRequest grab the PlasmaStore.cc mutex before calling ProcessCreateRequests, which is thread safe;
- However, ProcessCreateRequests also schedule an asynchronous ProcessCreateRequests onto the PlasmaStore’s asio thread with a timer, which doesn’t grab the PlasmaStore.cc mutex

This issue is revealed by this crash, where `plasma::PlasmaStore::IsObjectSpillable()` is called from a different thread.
```
Reduce Progress.:   0%|          | 0/5000 [23:53<?, ?it/s](raylet) *** SIGSEGV received at time=1630457392 on cpu 10 ***
(raylet) PC: @     0x5629c576222a  (unknown)  plasma::ObjectStore::GetObject()
(raylet)     @     0x7f783cf9d3c0  1481612800  (unknown)
(raylet)     @     0x5629c575603f         48  plasma::PlasmaStore::IsObjectSpillable()
(raylet)     @     0x5629c56f393c        240  ray::raylet::LocalObjectManager::SpillObjectsOfSize()
(raylet)     @     0x5629c56f3daf         64  ray::raylet::LocalObjectManager::SpillObjectUptoMaxThroughput()
(raylet)     @     0x5629c586f946        112  boost::asio::detail::completion_handler<>::do_complete()
(raylet)     @     0x5629c5d0d348        112  boost::asio::detail::scheduler::do_run_one()
(raylet)     @     0x5629c5d0e721        160  boost::asio::detail::scheduler::run()
(raylet)     @     0x5629c5d106b0         64  boost::asio::io_context::run()
(raylet)     @     0x5629c55d35d9       1456  main
(raylet)     @     0x7f783ca670b3  (unknown)  __libc_start_main
```
https://beta.anyscale.com/o/anyscale-internal/projects/prj_2xR6uT6t7jJuu1aCwWMsle/clu[…]wRSu9Wep769HhJPTSWgL?command-history-section=command_history

we also switch to use absl::mutex with annotations

### CODE REVIEW

**Comment by stephanie-wang** [comment]
Hmm I'm not sure if I understand the data race here. I thought the plasma store event loop only has one thread?

**Comment by scv119** [comment]
ah the [PlasmaStore::IsObjectSpillable](
https://github.com/ray-project/ray/blob/master/src/ray/object_manager/plasma/store_runner.cc#L118) is called from a different thread from LocalObjectManager

**Review by rkooo567 — APPROVED** [review]

**Comment by scv119** [comment]
@stephanie-wang suggested we should use absl::mutex and lock annotation to prevent this from happening again, which is a great idea.  I'm thinking I can do that once https://github.com/ray-project/ray/pull/18064 which makes the merge work a bit easier.

**Review by stephanie-wang — APPROVED** [review]
Thanks for finding this! Yes, we can do the annotations in a follow-up.

**Comment by scv119** [comment]
Working on some tests

File: src/ray/object_manager/plasma/store.cc [hunk 1]
```diff
@@ -648,6 +648,7 @@ void PlasmaStore::DoAccept() {
 }
 
 void PlasmaStore::ProcessCreateRequests() {
+  std::lock_guard<std::recursive_mutex> guard(mutex_);
```
Comment (ericl) [inline src/ray/object_manager/plasma/store.cc:651]: Hmm would this have been caught with absl annotations?

Comment (ericl) [reply src/ray/object_manager/plasma/store.cc:651]: We might want to switch to absl locks to enable this during refactoring.

**Review by scv119 — COMMENTED** [review]
Switch to absl::mutex and add mutex annotations.

File: src/ray/object_manager/plasma/store.cc [hunk 2]
```diff
@@ -89,11 +89,15 @@ PlasmaStore::PlasmaStore(instrumented_io_context &main_service, IAllocator &allo
...
+          // absl can't check thread safety for lambda
+          [this]() ABSL_NO_THREAD_SAFETY_ANALYSIS { return GetDebugDump(); }),
+      total_consumed_bytes_(0),
       get_request_queue_(io_context_, object_lifecycle_mgr_,
-                         [this](const ObjectID &object_id, const auto &request) {
-                           this->AddToClientObjectIds(object_id, request->client);
-                         },
+                         // absl failed to check thread safety for lambda
+                         [this](const ObjectID &object_id, const auto &request)
+                             ABSL_NO_THREAD_SAFETY_ANALYSIS {
```
Comment (scv119) [inline src/ray/object_manager/plasma/store.cc:98]: cc @mwtian I can't work around the annotation with lambda and my conclusion is absl can't analyze thread safety for lambdas (which makes sense)

File: src/ray/object_manager/plasma/store.h [hunk 3]
```diff
@@ -68,6 +70,42 @@ class PlasmaStore {
   /// Stop this store.
   void Stop();
 
+  /// Return true if the given object id has only one reference.
```
Comment (scv119) [inline src/ray/object_manager/plasma/store.h:73]: mostly just move a lot of functions as private and add annotations.

File: src/ray/object_manager/plasma/store.cc (hunk 2 above)
Comment (mwtian) [reply src/ray/object_manager/plasma/store.cc:98]: Yes I remember there were issues with lambda and thread safety analysis. Maybe higher versions of clang can help but not 100% sure.

File: src/ray/object_manager/plasma/store.cc [hunk 4]
```diff
@@ -345,9 +347,10 @@ Status PlasmaStore::ProcessMessage(const std::shared_ptr<Client> &client,
     const auto &request = flatbuffers::GetRoot<fb::PlasmaCreateRequest>(input);
     const size_t object_size = request->data_size() + request->metadata_size();
 
-    auto handle_create = [this, client, message](bool fallback_allocator,
-                                                 PlasmaObject *result,
-                                                 bool *spilling_required) {
+    // absl failed analyze mutex safety for lambda
```
Comment (ericl) [inline src/ray/object_manager/plasma/store.cc:350]: Can we do something like add an assert held statement for the lock? The might satisfy absl

**Review by rkooo567 — APPROVED** [review]

File: src/ray/object_manager/plasma/store.cc (hunk 4 above)
Comment (ericl) [reply src/ray/object_manager/plasma/store.cc:354]: Is absl smart enough that you can enable analysis with the assert in?

Comment (scv119) [reply src/ray/object_manager/plasma/store.cc:354]: :( sadly no
